### PR TITLE
REF: use cached inferred_type when calling lib.infer_dtype(index)

### PR DESCRIPTION
**Pull request by jbrockmendel**
Among other things, this will let us avoid a couple of ugly calls in Series

```
        if isinstance(key, Index):
            key_type = key.inferred_type
        else:
            key_type = lib.infer_dtype(key, skipna=False)
```

Cleanup of nearby EA-handling code

### CODE REVIEW

File: pandas/_libs/lib.pyx [hunk 1]
```diff
@@ -1291,44 +1291,49 @@ def infer_dtype(value: object, skipna: bool = True) -> str:
...
+
     else:
         if not isinstance(value, list):
             value = list(value)
-        from pandas.core.dtypes.cast import (
-            construct_1d_object_array_from_listlike)
-        values = construct_1d_object_array_from_listlike(value)
+
+        # See also: construct_1d_object_array_from_listlike
+        values = np.empty(len(value), dtype=object)
```
Comment (jreback) [inline pandas/_libs/lib.pyx:1317]: why did you change from using the function for this?

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:1317]: trying to untangle dependency structure

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx:1317]: can revert, this isnt related to the rest of the PR

Comment (jreback) [reply pandas/_libs/lib.pyx:1317]: that function exists for a reason, to avoid constantly repeating the 2 lines you added. pls revert.

Comment (jbrockmendel) [reply pandas/_libs/lib.pyx]: updated + green

**Comment by jreback** [comment]
lgtm. can you merge master and ping on green.

**Comment by jbrockmendel** [comment]
ping
